text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,274 @@
+
+#### Metadata descriptor
+
+Values for:
+
+1. `u8` metadata shortening protocol version, 
+2. SCALE-encoded `ExtrinsicMetadata`,
+3. SCALE-encoded `spec_version` `String`,
+4. SCALE-encoded `spec_name` `String`,
+5. `u16` base58 prefix,
+6. `u8` decimals value or `0u8` if no units are defined,
+7. SCALE-encoded `tokenSymbol` `String` defined on chain to identify the name of currency (available for example through `system.properties()` RPC call) or empty string if no base units are defined,
+
+```
+struct MetadataDescriptor { // really a scale-encoded enum, thus first field is enum value - only 0x01 currently supported.
+  protocol_version: u8,
+  extrinsic_metadata: Vec<u8>, // SCALE from `ExtrinsicMetadata
+  spec_version: Vec<u8>, // SCALE form `String`
+  spec_name: Vec<u8>, // SCALE from `String`
+  base58_prefix: u16,
+  decimals: u8,
+  token_symbol: Vec<u8>, // SCALE from `String`
+}
+```
+
+constitute metadata descriptor. This is minimal information that is, together with (shortened) types registry, sufficient to decode any signable transaction.
+
+#### Merkle tree
+
+A **Complete Binary Merkle Tree** (**CBMT**) is proposed as digest structure.
+
+Every node of the proposed tree has a 32-bit value.
+
+A terminal node of the tree we call **leaf**. Its value is input for digest.
+
+The top node of the tree we call **root**.
+
+All node values for non-leave nodes are not terminal are computed through non-commutative **merge** procedure of child nodes.
+
+In CBMT, all layers must be populated, except for the last one, that must have complete filling from the left.
+
+Nodes are numbered top-down and left-to-right starting with 0 at the top of tree.
+
+```
+Example 8-node tree
+
+        0
+     /     \
+    1       2
+   / \     / \
+  3   4   5   6
+ / \
+7   8
+
+Nodes 4, 5, 6, 7, 8 are leaves
+Node 0 is root
+
+```
+
+### General flow
+
+1. The metadata is converted into lean modular form (vector of chunks)
+2. A Merkle tree is constructed from the metadata chunks
+3. A root of tree (as a left element) is merged with Metadata Descriptor (as a right element)
+4. Resulting value is a constant to be included in `additionalSigned` to prove that the metadata seen by cold device is genuine
+
+### Metadata modularization
+
+1. Types registry is stripped from `docs` fields.
+2. Types records are separated into chunks, with enum variants being individual chunks differing by variant index; each chunk consisting of `id` (same as in full metadata registry) and SCALE-encoded 'Type' description (reduced to 1-variant enum for enum variants). Enums with 0 variants are treated as regular types.
+3. Chunks are sorted by `id` in ascending order; chunks with same `id` are sorted by enum variant index in ascending order.
```

Comment:
We clearly should not mention the `docs` field, as I assume that this isn't present in the "shortened metadata" representation. In general this algorithm does not make that much difference, if you don't know the types.

Reply:
Types registry comes directly from full, non-shortened metadata, that has all fields, and this (1) is the stage where it loses `docs`, not earlier. As we describe algorithm here, this is an essential step. The only other way to describe this would be copying all other fields to new structure, but that is just longer description of the same step.

Also, there is ambiguity about docs field missing altogether vs being blank that would probably surface across implementations; this statement should resolve it.